### ISSUE TITLE
How does the git pull upstream work?

### ISSUE THREAD

**Issue by scop22** [jwasham/coding-interview-university#1392]
Hi, thanks for this amazing resource.

Sorry if this is a really newb question, but I am trying to understand how the 'git pull upstream main' works. Will it overwrite the checkboxes I have checked on my local fork?

**Comment by scop22** [comment]
Thanks a lot for the reply. 👍 

I have followed the instructions from the git section at https://github.com/jwasham/coding-interview-university#how-to-use-it.

I have done the following steps:

- Fork the repo to my own github repo.

Then:

``` 
git clone https://github.com/<YOUR_GITHUB_USERNAME>/coding-interview-university.git
cd coding-interview-university
git remote add upstream https://github.com/jwasham/coding-interview-university.git
git remote set-url --push upstream DISABLE  # so that you don't push your personal progress back to the original repo 
```

But I *haven't* done these steps below yet, because I am worried that doing "git pull upstream main" will overwrite any changes I have made locally (that is, when I start checking checkboxes as I work through the README).

```
git commit -am "Marked personal progress"
git pull upstream main  # keep your fork up-to-date with changes from the original repo 
```

It may be that I am just misunderstanding how it works, but I just want to be sure before I begin. My current understanding is that `git commit` will upload my changes to my GitHub repo, but `git pull` will overwrite the changes on my local cloned copy...

Thanks for any further explanation anyone can provide.

**Comment by Weav3r** [comment]
> @scop22 As I understand it, it wouldn't automatically overwrite changes, it's most likely going to attempt to pull changes down, then create a merge conflict warning

You are right

**Comment by jwasham** [comment]
Hi there,
If anyone has a way to make this process better, please update the documentation and create a pull request. 
I appreciate it.

**Comment by Weav3r** [comment]
This is going to be a long reply that warrants a tldr.
TLDR; No. It will either succeed or wouldn't if you have any changes made. Regarding the latter, you'll be asked to do any of the following; commit, stash  or resolve conflicts (if any).

Long answer:
Lets explain the commands as briefly as possible.

- Fork the repo to my own github repo
Creates a copy of the original repo to your Github repos (referencing the original). The repo is on Github servers not your local computer.

`git clone https://github.com/<YOUR_GITHUB_USERNAME>/coding-interview-university.git`
Downloads a the repo you forked onto you local computer. In doing so, a new folder/directory is created from the location where you ran the command with the repo's name; in this case **coding-interview-university**

`cd coding-interview-university`
Navigate to the folder (repo) that was just created on local computer. To work on the files in the repo you need to access it's content and that's all this step is enabling us to do.

At this point, you can modify the files in the repo on your local PC and save (Ctrl+S or Cmd+S) them. However, saving the files and committing them mean different things when it comes to git repos. To oversimplify, think of it like this; the former saves the changes but those changes cannot be recovered if the files get overwritten or corrupted while the latter saves the changes permanently (kind of-only on your local PC) and can be recovered if any of the files get overwritten or corrupted. So most of the time you'll be _saving and committing_ once you're done with your changes. But it is only local to your PC. To save it to the cloud (Github in this case), you need to upload (push) your **commits** to your repo stored on Github.

``
